src/python/WMComponent/ArchiveDataReporter/ArchiveDataPoller.py

```diff
@@ -11,64 +11,68 @@
 from WMCore.Services.WMArchiver.WMArchiver import WMArchiver
 from WMCore.Services.FWJRDB.FWJRDBAPI import FWJRDBAPI
 
+
 class ArchiveDataPoller(BaseWorkerThread):
     """
     Gather fwjr data and update to archiver\
     """
-    
+
     def __init__(self, config):
         """
         initialize properties specified from config
         """
         BaseWorkerThread.__init__(self)
         self.config = config
-                         
+
     def setup(self, parameters):
         """
         set db connection(couchdb, wmbs) to prepare to gather information
         """
         baseURL = self.config.JobStateMachine.couchurl
         dbname = "%s/fwjrs" % getattr(self.config.JobStateMachine, "couchDBName")
-         
+
         self.fwjrAPI = FWJRDBAPI(baseURL, dbname)
         self.wmarchiver = WMArchiver(self.config.ArchiveDataReporter.WMArchiverURL)
         self.numDocsRetrievePerPolling = getattr(self.config.ArchiveDataReporter, "numDocsRetrievePerPolling", 1000)
         self.numDocsUploadPerCall = getattr(self.config.ArchiveDataReporter, "numDocsUploadPerCall", 200)
-        
+
     def algorithm(self, parameters):
         """
         get information from wmbs, workqueue and local couch
         """
         try:
-            logging.info("Getting not archived data info from FWRJ db...")
-            data = self.fwjrAPI.getFWJRByArchiveStatus('ready', limit=1000)['rows']
-            
+            data = self.fwjrAPI.getFWJRByArchiveStatus('ready', limit=self.numDocsRetrievePerPolling)['rows']
+            logging.info("Found %i not archived documents from FWRJ db to upload to WMArchive.", len(data))
+
             for slicedData in grouper(data, self.numDocsUploadPerCall):
                 jobIDs = []
                 archiverDocs = []
                 for job in slicedData:
                     doc = createArchiverDoc(job)
                     archiverDocs.append(doc)
                     jobIDs.append(job["id"])
-                    
+
+                # probably useless... check whether we have duplicate docs to upload to WMArchive
+                duplicateIDs = set([x for x in jobIDs if jobIDs.count(x) > 1])
+                if duplicateIDs:
+                    logging.warn("Removing these duplicate entries before uploading them %s", duplicateIDs)
+                    archiverDocs = list(set(archiverDocs))
+
```

Comment:
On  0, Alan Malta Rodrigues <notifications@github.com> wrote:
> >              for slicedData in grouper(data, self.numDocsUploadPerCall):
> >                  jobIDs = []
> >                  archiverDocs = []
> >                  for job in slicedData:
> >                      doc = createArchiverDoc(job)
> >                      archiverDocs.append(doc)
> >                      jobIDs.append(job["id"])
> > -                    
> > +
> > +                # probably useless... check whether we have duplicate docs to upload to WMArchive
> > +                duplicateIDs = set([x for x in jobIDs if jobIDs.count(x) > 1])
> > +                if duplicateIDs:
> > +                    logging.warn("Removing these duplicate entries before uploading them %s", duplicateIDs)
> > +                    archiverDocs = list(set(archiverDocs))
> > +
> 
> > Seangchan is correct, we may only have duplicates if in first call WMArchive fails and then agent make another call with duplicated docs.
> 
> Which, on the client/agent side, there is no way to distinguish. Right?

On a client side you'll get after first try the status from the wmarchive.
If status is fail you know that you need to try again. But, yes, you don't know
which docs should be re-tried. We always said that client will "just retry
entire set". The WMArchive job is not to insert documents which are already
in archive, so from the client side it will be no difference if client
will sent 1, 2 or 10 times.